src/communication/communicationmanager.cpp

```diff
@@ -0,0 +1,229 @@
+    SHA256_Final(checksum, &sha256);
+}
+
+namespace aos::mp::communication {
+
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+
+aos::Error CommunicationManager::Init(const aos::mp::config::Config& cfg, TransportItf& transport,
+    aos::mp::iamclient::CertProviderItf* certProvider, aos::cryptoutils::CertLoaderItf* certLoader,
+    aos::crypto::x509::ProviderItf* cryptoProvider)
+{
+    LOG_DBG() << "Init communication manager";
+
+    mTransport      = &transport;
+    mCertProvider   = certProvider;
+    mCertLoader     = certLoader;
+    mCryptoProvider = cryptoProvider;
+    mCfg            = &cfg;
+
+    mThread = std::thread(&CommunicationManager::Run, this);
+
+    return aos::ErrorEnum::eNone;
+}
+
+std::shared_ptr<CommChannelItf> CommunicationManager::CreateChannel(
+    int port, aos::mp::iamclient::CertProviderItf* certProvider, const std::string& certStorage)
+{
+    auto chan = std::make_shared<CommunicationChannel>(port, this);
+
+    if (certProvider == nullptr) {
+        LOG_DBG() << "Create open channel";
+
+        // auto openchannel = std::make_unique<OpenChannel>(chan.get(), port);
+
+        mChannels[port] = chan;
+
+        return chan;
+    }
+
+    LOG_DBG() << "Create secure channel: port=" << port << ", certStorage=" << certStorage.c_str();
+
+    auto securechannel = std::make_shared<SecureChannel>(
+        *mCfg, *chan, *certProvider, *mCertLoader, *mCryptoProvider, port, certStorage);
+
+    mChannels[port] = std::move(chan);
+
+    return securechannel;
+}
+
+aos::Error CommunicationManager::Connect()
+{
+    {
+        std::lock_guard<std::mutex> lock(mMutex);
+
+        if (mIsConnected) {
+            return aos::ErrorEnum::eNone;
+        }
+
+        LOG_DBG() << "Connect communication manager";
+
+        auto err = mTransport->Connect();
+        if (!err.IsNone()) {
+            return err;
+        }
+
+        mIsConnected = true;
+    }
+
+    mCondVar.notify_all();
+
+    return aos::ErrorEnum::eNone;
+}
+
+aos::Error CommunicationManager::Read([[maybe_unused]] std::vector<uint8_t>& message)
+{
+    return aos::ErrorEnum::eNotSupported;
+}
+
+aos::Error CommunicationManager::Write(std::vector<uint8_t> message)
+{
+    std::unique_lock<std::mutex> lock(mMutex);
+    mCondVar.wait_for(lock, cConnectionTimeout, [this]() { return mIsConnected; });
+
+    if (!mIsConnected) {
+        return aos::ErrorEnum::eTimeout;
+    }
+
+    return mTransport->Write(std::move(message));
+}
+
+aos::Error CommunicationManager::Close()
+{
+    if (mShutdown) {
+        return aos::ErrorEnum::eNone;
+    }
+
+    LOG_DBG() << "Close communication manager";
+
+    mShutdown = true;
+
+    aos::Error err;
+
+    {
+        std::lock_guard<std::mutex> lock(mMutex);
```

Comment:
Change over the all sources as following:

std::lock_guard<std::mutex> lock(mMutex) - > std::lock_guard lock {mMutex};
std::unique_lock<std::mutex> lock(mMutex) -> std::unique_lock lock {mMutex};